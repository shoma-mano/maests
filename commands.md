# Measts Commands

# Measts Commands

## Table of Contents

- [Initialize Flow](#initialize-flow)
- [Taps and Presses](#taps-and-presses)
- [Swipes and Scrolls](#swipes-and-scrolls)
- [Text Input](#text-input)
- [Navigation and Links](#navigation-and-links)
- [Assertions](#assertions)
- [Waiting](#waiting)
- [Keyboard Actions](#keyboard-actions)
- [Device Volume Controls](#device-volume-controls)
- [Miscellaneous Actions](#miscellaneous-actions)
- [Repeat Actions](#repeat-actions)

### Initialize Flow

- **Launch the Application**
```typescript
    M.initFlow({ appId: "com.example.app" })
```

- **Clear Application State**
```typescript
    M.clearState({ appId: "com.example.app" })
```

- **Clear Keychain**
```typescript
    M.clearKeychain()
```

### Taps and Presses**

Taps and Presses have optional properties available for configuring various Maestro commands, designed to enhance the flexibility and customization of user interactions. The properties are described below.

#### Properties

### TapProps

The `TapProps` interface defines optional properties for customizing tap actions in Maestro commands, such as `tapOn`, `tapOnText`, and `tapOnPoint`.

| Property                | Type     | Default | Description                                                                                                  |
|-------------------------|----------|---------|--------------------------------------------------------------------------------------------------------------|
| `retryTapIfNoChange`    | boolean  | `true`  | If set to `true`, the tap action will be retried if no change is detected on the first attempt.              |
| `repeat`                | number   | -       | Specifies the number of times to repeat the tap action.                                                      |
| `waitToSettleTimeoutMs` | number   | -       | Time in milliseconds to wait for the element to settle after tapping, useful for handling UI transition time. |

#### Example Usage:
```typescript
M.tapOn("submitButton", { retryTapIfNoChange: false, repeat: 2, waitToSettleTimeoutMs: 500 });

---

### WaitProps

```markdown
### WaitProps

The `WaitProps` interface provides optional properties for configuring wait behavior in specific commands, such as `waitForAndTapOn`, `waitUntilVisible`, and `waitUntilNotVisible`.

| Property   | Type    | Default | Description                                                                                     |
|------------|---------|---------|-------------------------------------------------------------------------------------------------|
| `maxWait`  | number  | 5000    | The maximum time (in milliseconds) to wait for an element or condition before continuing.       |

#### Example Usage:
```typescript
M.waitForAndTapOn("startButton", { maxWait: 3000 });


---

### PointProps

```markdown
### PointProps

The `PointProps` interface specifies coordinates for point-based tap or long-press actions, enabling precise control of on-screen actions.

| Property | Type                | Description                                                                      |
|----------|----------------------|----------------------------------------------------------------------------------|
| `x`      | `number | string`   | The x-coordinate for the point to tap on. Can be an absolute number or percentage.|
| `y`      | `number | string`   | The y-coordinate for the point to tap on. Can be an absolute number or percentage.|

#### Example Usage:
```typescript
M.tapOnPoint({ x: "50%", y: "50%" }, { waitToSettleTimeoutMs: 200 });


- **Tap on Element by ID**

```typescript
    M.tapOn("submitButton", { retryTapIfNoChange: true, repeat: 2, waitToSettleTimeoutMs: 500 })
```
- **Tap on Visible Text**
```typescript
    M.tapOnText("Submit", { retryTapIfNoChange: true })
```
- **Tap on Point**
```typescript
    M.tapOnPoint({ x: "50%", y: "50%" }, { waitToSettleTimeoutMs: 200 })
```
- **Wait for Element to Appear and Tap**
```typescript
    M.waitForAndTapOn("submitButton", 3000)
```
- **Long Press on Element by ID**
```typescript
    M.longPressOn("submitButton")
```
- **Long Press on Text**
```typescript
    M.longPressOnText("Submit")
```

### Swipes and Scrolls

- **Swipe in a Direction**
```typescript
    M.swipeLeft()
    M.swipeRight()
    M.swipeUp()
    M.swipeDown()
```

- **Swipe from Point to Point**
```typescript
   M.swipe({ x: "10%", y: "20%" }, { x: "90%", y: "80%" })
```

- **Scroll Screen**
```typescript
    M.scroll()
```

- **Scroll Until Element is Visible**
```typescript
   M.scrollUntilVisible("scrollTarget")
```

### Text Input

- **Input Text into an Element**
```typescript
   M.inputText("Hello, World!", "textFieldId")
```

- **Input Random Name**
```typescript
    M.inputRandomName("nameFieldId")
```

- **Input Random Number**
```typescript
    M.inputRandomNumber("numberFieldId")
```

- **Input Random Email**
```typescript
    M.inputRandomEmail("emailFieldId")
```

- **Input Random Text**
```typescript
    M.inputRandomText("textFieldId")
```

- **Input Random Text**
```typescript
    M.inputRandomText("textFieldId")
```

- **Erase Text**
```typescript
    M.eraseText(10, "textFieldId")
```

### Navigation and Links

- **Open a Link**
```typescript
    M.openLink("https://example.com")
```

- **Navigate to Path Using Deep Link Base**
```typescript
   M.navigate("/home")
```

### Assertions

- **Assert Element is Visible**
```typescript
  M.assertVisible("submitButton", true)
```

- **Assert Element is Not Visible**
```typescript
  M.assertNotVisible("submitButton")
```

### Waiting
- **Wait for Animation End**
```typescript
  M.waitForAnimationEnd(3000)
```

- **Wait Until Element is Visible**
```typescript
  M.waitUntilVisible("submitButton", 5000)
```

- **Wait Until Element is Not Visiblee**
```typescript
  M.waitUntilNotVisible("loadingSpinner", 5000)
```

- **Wait a Specified Number of Milliseconds**
```typescript
  M.wait(1000)
```

### Keyboard Actions

- **Hide Keyboard**
```typescript
  M.hideKeyboard()
```

- **Press Enter**
```typescript
  M.pressEnter()
```

- **Press Home Button**
```typescript
  M.pressHomeButton()
```

- **Press Lock Button**
```typescript
  M.pressLockButton()
```

- **Press Back Button**
```typescript
  M.back()
```

### Device Volume Controls

- **Increase Volume**
```typescript
    M.volumeUp()
```

- **Decrease Volume**
```typescript
   M.volumeDown()
```

### Miscellaneous Actions

- **Take Screenshot**
```typescript
    M.screenshot("my-screenshot.png")
```

- **Copy Text from an Element**
```typescript
    M.copyTextFrom("textElementId")
```

- **Assert Condition is True**
```typescript
    M.assertTrue("2 + 2 === 4")
```

### Repeat Actions

- **Repeat Actions a Specified Number of Times**
```typescript
    M.repeat(3, () => {
    M.tapOn("submitButton");
    M.wait(500)
});
```

- **Repeat Actions While Element is Visibles**
```typescript
    M.repeatWhileVisible("loadingSpinner", () => {
    M.wait(1000)
});
```
- **Repeat Actions While Element is Not Visible**
```typescript
    M.repeatWhileNotVisible("submitButton", () => {
    M.wait(1000);
});
```






































