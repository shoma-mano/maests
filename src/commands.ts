// nested commands
let isNested = false;
let nestedCommands = "";
const handleNest = (func: () => any) => {
  isNested = true;
  func();
  isNested = false;
  const out = nestedCommands;
  nestedCommands = "";
  return out;
};

export let out = "";
export const resetOut = () => {
  out = "";
};
const addOut = (command: string) => {
  if (isNested) nestedCommands += command;
  else out += command;
};

const envAppId = process.env["appId"];
export const MaestroTranslators = {
  /**
   * Should be called at the start of every test flow.
   * In the config object, you can define the appId to use.
   */
  initFlow: ({ appId }: { appId?: string } = {}) => {
    addOut(`appId: ${appId ?? envAppId}\n---\n`);
  },
  /**
   * Launches the app.
   * @param appId The bundle id of your app. Falls back to the appId provided in maestro-ts.config.js.
   */
  launchApp: ({ appId }: { appId?: string } = {}) => {
    addOut("- launchApp:\n" + `    appId: "${appId ?? envAppId}"\n`);
  },
  /**
   * Clear the state of the current app or of the app with the given id.
   */
  clearState: ({ appId }: { appId?: string } = {}) => {
    if (appId) addOut(`- clearState: ${appId ?? envAppId}\n`);
    addOut("- clearState\n");
  },
  /**
   * Clear the entire keychain.
   */
  clearKeychain: () => {
    addOut("- clearKeychain\n");
  },
  /**
   * Tap on an element with the given testId.
   */
  tapOn: (id: string) => {
    addOut(`- tapOn:\n    id: "${id}"\n`);
  },
  /**
   * Tap on a text visible on screen.
   */
  tapOnText: (text) => {
    addOut(`- tapOn: ${text}\n`);
  },
  /**
   * Tap on the given point.
   * Can either take numbers for dips or strings for percentages.
   */
  tapOnPoint: ({ x, y }) => {
    addOut(`- tapOn:\n    point: ${x},${y}\n`);
  },
  /**
   * Wait for testId to appear and the tap on an element with the given testId.
   */
  waitForAndtapOn: (id, maxWait) => {
    addOut(
      "- extendedWaitUntil:\n" +
        "    visible:\n" +
        `        id: "${id}"\n` +
        `    timeout: ${maxWait}\n`
    );
    addOut(`- tapOn:\n    id: "${id}"\n`);
  },
  /**
   * Long press on an element with the given testId.
   */
  longPressOn: (id) => {
    addOut(`- longPressOn:\n    id: "${id}"\n`);
  },
  /**
   * Long press on the given point.
   */
  longPressOnPoint: ({ x, y }) => {
    addOut(`- longPressOn:\n    point: ${x}, ${y}\n`);
  },
  /**
   * Long press on an element with the given text.
   */
  longPressOnText: (text) => {
    addOut(`- longPressOn: ${text}\n`);
  },
  /**
   * Swipe left from center.
   */
  swipeLeft: () => {
    addOut("- swipe:\n" + "    direction: LEFT\n" + "    duration: 400\n");
  },
  /**
   * Swipe right from center.
   */
  swipeRight: () => {
    addOut("- swipe:\n" + "    direction: RIGHT\n" + "    duration: 400\n");
  },
  /**
   * Swipe down from center.
   */
  swipeDown: () => {
    addOut("- swipe:\n" + "    direction: DOWN\n" + "    duration: 400\n");
  },
  /**
   * Swipe up from center.
   */
  swipeUp: () => {
    addOut("- swipe:\n" + "    direction: UP\n" + "    duration: 400\n");
  },
  /**
   * Swipe from a start to an end point. Use percentages or dips.
   */
  swipe: (start: { x: number; y: number }, end: { x: number; y: number }) => {
    addOut(
      `- swipe:\n    start: ${start.x}, ${start.y}\n    end: ${end.x}, ${end.y}\n`
    );
  },
  /**
   * Input a text into the currently focused input or the input with the given testId.
   */
  inputText: (text: string, id?: string) => {
    if (!id) addOut(`- inputText: ${text}\n`);
    addOut(`- tapOn:\n    id: "${id}"\n- inputText: ${text}\n`);
  },
  /**
   * Input random name into focused input or the one with given testId.
   */
  inputRandomName: (id?: string) => {
    if (!id) addOut(`- inputRandomPersonName\n`);
    addOut(`- tapOn:\n    id: "${id}"\n- inputRandomPersonName\n`);
  },
  /**
   * Input random number into focused input or the one with given testId.
   */
  inputRandomNumber: (id?: string) => {
    if (!id) addOut(`- inputRandomNumber\n`);
    addOut(`- tapOn:\n    id: "${id}"\n- inputRandomNumber\n`);
  },
  /**
   * Copies text of an element with the given testId.
   */
  copyTextFrom: (id: string) => {
    addOut(`- copyTextFrom:\n    id: "${id}"\n`);
  },
  /**
   * Input random email into focused input or the one with given testId.
   */
  inputRandomEmail: (id?: string) => {
    if (!id) addOut(`- inputRandomEmail\n`);
    addOut(`- tapOn:\n    id: "${id}"\n- inputRandomEmail\n`);
  },
  /**
   * Input random text into focused input or the one with given testId.
   */
  inputRandomText: (id?: string) => {
    if (!id) addOut(`- inputRandomText\n`);
    addOut(`- tapOn:\n    id: "${id}"\n- inputRandomText\n`);
  },
  /**
   * Erase a number of characters from the focused input or the input with the given testId.
   */
  eraseText: (chars: number, id?: string) => {
    if (!id) addOut(`- eraseText: ${chars ?? 50}\n`);
    addOut(`- tapOn:\n    id: "${id}"\n- eraseText: ${chars ?? 50}\n`);
  },
  /**
   * Open a url / deepLink.
   */
  openLink: (url) => {
    addOut(`- openLink: ${url}\n`);
  },
  /**
   * Use the configured deepLinkBase or appId to navigate to the given path.
   * Only works if deepLinking is set up correctly.
   */
  navigate: (path) => {
    addOut(`- openLink: ${process.env["deepLinkBase"]}${path}\n`);
  },
  /**
   * Assert an element with the given testId is visible.
   * @param enabled Whether the view should also be enabled.
   */
  assertVisible: (id, enabled) => {
    if (enabled)
      addOut(`- assertVisible:\n    id: "${id}"\n    enabled: true\n`);
    addOut(`- assertVisible:\n    id: "${id}"\n`);
  },
  /**
   * Assert the element with the given testId is not visible.
   */
  assertNotVisible: (id) => {
    addOut(`- assertNotVisible:\n    id: "${id}"\n`);
  },
  /**
   * Scroll down.
   */
  scroll: () => {
    addOut(`- scroll\n`);
  },
  /**
   * Scroll until the element with the given testId is visible.
   */
  scrollUntilVisible: (id) => {
    addOut(`- scrollUntilVisible:\n    element:\n      id: "${id}"\n`);
  },
  /**
   * Wait a max of n ms or until the current animation has ended.
   */
  waitForAnimationEnd: (continueAfter?: boolean) => {
    if (!continueAfter) {
      addOut("- waitForAnimationToEnd\n");
    }
    addOut(`- waitForAnimationToEnd:\n    timeout: ${continueAfter}\n`);
  },
  /**
   * Wait a max of milliseconds until the element with the given testId is visible.
   */
  waitUntilVisible: (id, maxWait) => {
    addOut(
      "- extendedWaitUntil:\n" +
        "    visible:\n" +
        `        id: "${id}"\n` +
        `    timeout: ${maxWait ?? 5000}\n`
    );
  },
  /**
   * Wait a max of milliseconds until the element with the given testId is no longer visible.
   */
  waitUntilNotVisible: (id, maxWait) => {
    addOut(
      "- extendedWaitUntil:\n" +
        "    notVisible:\n" +
        `        id: "${id}"\n` +
        `    timeout: ${maxWait ?? 5000}\n`
    );
  },
  /**
   * Wait a number of milliseconds.
   * This is an anti-pattern, try to fall back to other waiting methods if possible.
   */
  wait: (ms) => {
    addOut(
      "- swipe:\n" +
        "    start: -1, -1\n" +
        "    end: -1, -100\n" +
        `    duration: ${ms}\n`
    );
  },
  /**
   * Dismiss the software keyboard.
   */
  hideKeyboard: () => {
    addOut("- hideKeyboard\n");
  },
  /**
   * Take a screenshot and store at the path with the given name.
   */
  screenshot: (fileName) => {
    addOut(`- takeScreenshot: ${fileName}\n`);
  },
  /**
   * Press the enter key on the software keyboard.
   */
  pressEnter: () => {
    addOut("- pressKey: Enter\n");
  },
  /**
   * Press the home button.
   */
  pressHomeButton: () => {
    addOut("- pressKey: Home\n");
  },
  /**
   * Press the lock button.
   */
  pressLockButton: () => {
    addOut("- pressKey: Lock\n");
  },
  /**
   * Press android back button.
   */
  back: () => {
    addOut("- pressKey: back\n");
  },
  /**
   * Decrease device volume.
   */
  volumeDown: () => {
    addOut("- pressKey: volume down\n");
  },
  /**
   * Increase device volume.
   */
  volumeUp: () => {
    addOut("- pressKey: volume up\n");
  },
  /**
   * Stop the current app or the one with the given appId.
   */
  stopApp: ({ appId }: { appId?: string } = {}) => {
    appId = appId ?? envAppId;
    if (appId) addOut(`- stopApp: ${appId}\n`);
    addOut("- stopApp\n");
  },
  /**
   * Repeats the given actions a given number of times.
   */
  repeat: (times, func) => {
    const out = handleNest(func);
    const commands = `- repeat:
    times: ${times}
    commands:
        ${out.replace(/\n(?=.*[\n])/g, "\n        ")}`;
    addOut(commands);
  },
  /**
   * Repeats the given actions while the element with the given testId is visible.
   */
  repeatWhileVisible: (id, func) => {
    const out = handleNest(func);
    addOut(`- repeat:
    while:
        visible:
            id: "${id}"
    commands:
        ${out.replace(/\n(?=.*[\n])/g, "\n        ")}`);
  },
  /**
   * Repeats the given actions while the element with the given testId is not visible.
   */
  repeatWhileNotVisible: (id, func) => {
    const out = handleNest(func);
    addOut(`- repeat:
    while:
        notVisible:
            id: "${id}"
    commands:
        ${out.replace(/\n(?=.*[\n])/g, "\n        ")}`);
  },
  /**
   * Insert inline yaml code. Good for specialized commands.
   */
  yaml: (yaml: string) => `${yaml}\n`,
  /**
   * Check if a condition is true.
   */
  assertTrue: (condition: string) => {
    addOut(`- assertTrue: ${condition}\n`);
  },
};

export { MaestroTranslators as M };
export { writeYaml } from "./write-yaml";
