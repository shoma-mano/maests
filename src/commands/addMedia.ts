import { stringify } from "yaml";
import { addOut, getOut } from "../out";

// List of MIME types supported by Maestro for adding media to the device's gallery
const allowedMimeTypes = ["image/png", "image/jpeg", "image/jpg", "image/gif", "video/mp4"];

/**
 * Adds a media file to the device's gallery, making it accessible for use in application flows.
 * Supports both Android and iOS platforms.
 * 
 * @param {string} mimeType - The MIME type of the media file to add.
 *                            Must be one of the following supported types: 
 *                            "image/png", "image/jpeg", "image/jpg", "image/gif", "video/mp4".
 * @param {string} filePath - The relative path to the media file.
 * 
 * @throws {Error} Throws an error if the provided MIME type is not in the allowed list.
 * 
 * @example
 * addMedia("image/png", "../assets/test.png"); // Adds a PNG file located at the specified path
 */
export const addMedia = (mimeType: string, filePath: string) => {
  // Validate if the provided MIME type is in the list of allowed types
  if (!allowedMimeTypes.includes(mimeType.toLowerCase())) {
    throw new Error(`Unsupported MIME type. Allowed types are: ${allowedMimeTypes.join(", ")}`);
  }

  // Command structure for adding media to the gallery
  const commands = [
    {
      addMedia: [`${filePath}`]
    },
  ];
  addOut(stringify(commands));
};

// Tests for the addMedia function using Vitest
if (import.meta.vitest) {
  it("should match snapshot for addMedia with a valid mimeType and filePath", () => {
    addMedia('image/png', '../assets/test.png');
    expect(getOut()).toMatchInlineSnapshot(`
      "- addMedia:
          - ../assets/test.png
      "
    `);
  });

  it("should throw an error for an unsupported mimeType", () => {
    expect(() => addMedia("application/pdf", "../assets/test.pdf")).toThrowError(
      "Unsupported MIME type. Allowed types are: image/png, image/jpeg, image/jpg, image/gif, video/mp4"
    );
  });
}
