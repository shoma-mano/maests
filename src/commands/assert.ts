import { stringify } from "yaml";
import { addOut, getOut } from "../out";

type AssertProps = {
  text?: string;
  id?: string;
  enabled?: boolean;
  checked?: boolean;
  focused?: boolean;
  selected?: boolean;
};

export const assertVisible = (props: AssertProps) => {
  const commands = [{ assertVisible: { ...props } }];
  addOut(stringify(commands));
};

if (import.meta.vitest) {
  it("assertVisible", () => {
    assertVisible({ id: "com.android.systemui:id/battery" });
    expect(getOut()).toMatchInlineSnapshot(`
      "- assertVisible:
          id: com.android.systemui:id/battery
      "
    `);
  });
}

export const assertNotVisible = (props: AssertProps) => {
  const commands = [{ assertNotVisible: { ...props } }];
  addOut(stringify(commands));
};

if (import.meta.vitest) {
  it("assertNotVisible", () => {
    assertNotVisible({ id: "com.android.systemui:id/battery" });
    expect(getOut()).toMatchInlineSnapshot(`
      "- assertNotVisible:
          id: com.android.systemui:id/battery
      "
    `);
  });
}
