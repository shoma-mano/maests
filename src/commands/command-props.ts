export interface WaitProps {
  maxWait?: number;
}

export interface PointProps {
  x: number | string;
  y: number | string;
}
