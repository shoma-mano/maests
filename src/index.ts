#!/usr/bin/env node
import fs, { writeFileSync } from "fs";
import path, { dirname, join } from "path";
import createJiti from "jiti";
import dotenv from "dotenv";
import { consola } from "consola";
import { rewriteCode } from "./rewriteCode";
import { defineCommand, runMain } from "citty";
import { createYamlOutPath } from "./utils";
import { execSync } from "child_process";
import { getTsconfig } from "get-tsconfig";

const main = defineCommand({
  args: {
    path: {
      type: "positional",
      required: true,
    },
    device: {
      type: "string",
      alias: "d",
      description: "Device to run the test on",
    },
  },
  async run({ args }) {
    loadEnv();
    const cwd = process.cwd();

    // create temp file
    let flowPath = args.path;
    if (flowPath.startsWith(cwd)) {
      flowPath = flowPath.replace(`${cwd}/`, "");
    }
    const yamlOutPath = createYamlOutPath(flowPath);
    let code = fs.readFileSync(flowPath, "utf-8");
    code = rewriteCode({ code, yamlOutPath });
    const tempFilePath = join(cwd, `${flowPath.replace(".ts", ".temp.ts")}`);
    writeFileSync(tempFilePath, code);

    // create jiti instance
    const { config, path } = getTsconfig();
    const normalizedAlias = Object.fromEntries(
      Object.entries(config?.compilerOptions?.paths || {}).map(
        ([key, value]) => {
          const normalizedKey = key.replace("/*", "");
          const normalizedValue = value[0].replace("/*", "");
          return [normalizedKey, join(dirname(path), normalizedValue)];
        }
      )
    );
    const jiti = createJiti(cwd, {
      esmResolve: true,
      experimentalBun: true,
      alias: normalizedAlias,
    });

    // execute temp file
    try {
      await jiti(tempFilePath);
      consola.success(`Created Yaml to ${yamlOutPath} âœ”`);
    } catch (e) {
      console.error(e);
      fs.unlinkSync(tempFilePath);
      process.exit(1);
    }
    fs.unlinkSync(tempFilePath);

    // run maestro test
    const command = `maestro ${
      args.device ? `--device ${args.device}` : ""
    } test  ${yamlOutPath}`;
    try {
      execSync(command, {
        stdio: "inherit",
        env: process.env,
      });
    } catch (e) {
      console.error(e);
      process.exit(1);
    }
  },
});
if (!import.meta.vitest) runMain(main);

function loadEnv() {
  let currentPath = process.cwd();
  let dotEnvPath = "";
  while (currentPath !== "/") {
    const files = fs.readdirSync(currentPath);
    if (files.includes(".env")) {
      dotEnvPath = path.join(currentPath, ".env");
      break;
    }
    currentPath = path.join(currentPath, "../");
  }
  if (dotEnvPath) {
    consola.info(`Found .env file at ${dotEnvPath}`);
    dotenv.config({ path: dotEnvPath });
  }
}
