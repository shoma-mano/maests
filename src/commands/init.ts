import { handleNest, space, addOut, getOutput } from "./commands";

const envAppId = process.env["appId"];

export const initFlow = ({
  appId,
  onFlowStart,
}: { appId?: string; onFlowStart?: () => void } = {}) => {
  const appIdCommand = `appId: ${appId ?? envAppId}\n`;
  let commands = appIdCommand;
  if (onFlowStart) {
    const nested = handleNest(onFlowStart);
    const flowCommand = `onFlowStart:\n${nested.replaceAll(
      /\n/g,
      `${space}\n`
    )}`;
    commands += flowCommand;
  }
  const separator = "---\n";
  commands += separator;
  addOut(commands);
};

if (import.meta.vitest) {
  it("initFlow with appId", () => {
    initFlow({ appId: "testAppId" });
    expect(getOutput()).toMatchInlineSnapshot(`
        "appId: testAppId
        ---
        "
      `);
  });
}

export const launchApp = ({ appId }: { appId?: string } = {}) => {
  addOut(`- launchApp:\n    appId: "${appId ?? envAppId}"\n`);
};

if (import.meta.vitest) {
  it("launchApp with appId", () => {
    launchApp({ appId: "testAppId" });
    expect(getOutput()).toMatchInlineSnapshot(`
        "- launchApp:
            appId: "testAppId"
        "
      `);
  });
}

export const clearState = ({ appId }: { appId?: string } = {}) => {
  addOut(appId ? `- clearState: ${appId}\n` : "- clearState\n");
};

if (import.meta.vitest) {
  it("clearState with appId", () => {
    clearState({ appId: "testAppId" });
    expect(getOutput()).toMatchInlineSnapshot(`
        "- clearState: testAppId
        "
      `);
  });
}
