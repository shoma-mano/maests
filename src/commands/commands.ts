import { tapOn, tapOnPoint, tapOnText, waitForAndTapOn } from "./tap";
import { runFlow, runScript } from "./run";
import { clearState, initFlow, launchApp } from "./init";
import { repeat, repeatWhileVisible, repeatWhileNotVisible } from "./repeat";
import { assertNotVisible, assertVisible } from "./assert";
import { addOut } from "../out";
import { PointProps } from "../type";

// Utility function for indenting except last line break
export const indentExceptLastLineBreak = (str: string) => {
  return str.replace(/\n(?=.*[\n])/g, "\n        ");
};

export const space = "    ";

// Main translator functions
export const MaestroTranslators = {
  /**
   * Initializes the test flow with optional configuration.
   * @param config - Optional configuration with appId and other environment variables.
   */
  initFlow,
  /**
   * Launches the application with optional configuration settings.
   * @param config - Configuration options for appId, state clearing, keychain clearing, and app stopping.
   */
  launchApp,
  /**
   * Clears the state of the current app or the specified app by appId.
   * @param appId - Optional appId to clear state for a specific app.
   */
  clearState,

  runScript,

  runFlow,

  tapOn,

  tapOnText,

  tapOnPoint,

  waitForAndTapOn,

  clearKeychain: () => {
    addOut("- clearKeychain\n");
  },

  longPressOn: (id: string) => {
    addOut(`- longPressOn:\n    id: "${id}"\n`);
  },

  longPressOnPoint: (pointProps: PointProps) => {
    addOut(`- longPressOn:\n    point: ${pointProps.x},${pointProps.y}\n`);
  },

  longPressOnText: (text: string) => {
    addOut(`- longPressOn: ${text}\n`);
  },

  swipeLeft: () => addOut("- swipe:\n    direction: LEFT\n    duration: 400\n"),
  swipeRight: () =>
    addOut("- swipe:\n    direction: RIGHT\n    duration: 400\n"),
  swipeDown: () => addOut("- swipe:\n    direction: DOWN\n    duration: 400\n"),
  swipeUp: () => addOut("- swipe:\n    direction: UP\n    duration: 400\n"),

  swipe: (start: PointProps, end: PointProps) => {
    addOut(
      `- swipe:\n    start: ${start.x}, ${start.y}\n    end: ${end.x}, ${end.y}\n`
    );
  },

  inputText: (text: string, id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputText: ${text}\n`
        : `- inputText: ${text}\n`
    );
  },

  inputRandomName: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomPersonName\n`
        : `- inputRandomPersonName\n`
    );
  },

  inputRandomNumber: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomNumber\n`
        : `- inputRandomNumber\n`
    );
  },

  copyTextFrom: (id: string) => {
    addOut(`- copyTextFrom:\n    id: "${id}"\n`);
  },

  inputRandomEmail: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomEmail\n`
        : `- inputRandomEmail\n`
    );
  },

  inputRandomText: (id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- inputRandomText\n`
        : `- inputRandomText\n`
    );
  },

  eraseText: (chars: number, id?: string) => {
    addOut(
      id
        ? `- tapOn:\n    id: "${id}"\n- eraseText: ${chars ?? 50}\n`
        : `- eraseText: ${chars ?? 50}\n`
    );
  },

  openLink: (url: string) => {
    addOut(`- openLink: ${url}\n`);
  },

  navigate: (path: string) => {
    addOut(`- openLink: ${process.env["deepLinkBase"]}${path}\n`);
  },

  assertVisible,

  assertNotVisible,

  scroll: () => {
    addOut("- scroll\n");
  },

  scrollUntilVisible: (id: string) => {
    addOut(`- scrollUntilVisible:\n    element:\n      id: "${id}"\n`);
  },

  waitForAnimationEnd: (maxWait: number = 5000) => {
    addOut(
      maxWait
        ? `- waitForAnimationToEnd:\n    timeout: ${maxWait}\n`
        : "- waitForAnimationToEnd\n"
    );
  },

  waitUntilVisible: (id: string, maxWait: number) => {
    addOut(
      `- extendedWaitUntil:\n    visible:\n        id: "${id}"\n    timeout: ${
        maxWait ?? 5000
      }\n`
    );
  },

  waitUntilNotVisible: (id: string, maxWait: number) => {
    addOut(
      `- extendedWaitUntil:\n    notVisible:\n        id: "${id}"\n    timeout: ${
        maxWait ?? 5000
      }\n`
    );
  },

  wait: (ms: number) => {
    addOut(
      `- swipe:\n    start: -1, -1\n    end: -1, -100\n    duration: ${ms}\n`
    );
  },

  hideKeyboard: () => {
    addOut("- hideKeyboard\n");
  },

  screenshot: (fileName: string) => {
    addOut(`- takeScreenshot: ${fileName}\n`);
  },

  pressEnter: () => {
    addOut("- pressKey: Enter\n");
  },

  pressHomeButton: () => {
    addOut("- pressKey: Home\n");
  },

  pressLockButton: () => {
    addOut("- pressKey: Lock\n");
  },

  back: () => {
    addOut("- pressKey: back\n");
  },

  volumeDown: () => {
    addOut("- pressKey: volume down\n");
  },

  volumeUp: () => {
    addOut("- pressKey: volume up\n");
  },

  stopApp: ({ appId }: { appId?: string } = {}) => {
    addOut(appId ? `- stopApp: ${appId}\n` : "- stopApp\n");
  },

  repeat,

  repeatWhileVisible,

  repeatWhileNotVisible,

  yaml: (yaml: string) => `${yaml}\n`,

  assertTrue: (condition: string) => {
    addOut(`- assertTrue: ${condition}\n`);
  },
};

export { MaestroTranslators as M };
export { writeYaml } from "../write-yaml";

// utils for user
export const getOutput = (key: string) => "${output." + key + "}";

// runScript Types
declare global {
  namespace http {
    const get: (...args: any) => { body: string };
  }

  const json: <T extends any>(str: string) => T;
  const output: Record<string, string>;
}
