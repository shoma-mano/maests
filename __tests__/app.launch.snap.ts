import { describe, it, expect, beforeEach } from "vitest";
import { MaestroTranslators, resetOut, out } from "../src/commands";

const getOutput = () => {
  const currentOutput = out;
  resetOut(); // Reset output for clean snapshots in each test
  return currentOutput;
};

describe("MaestroTranslators Application Launch Commands Snapshots", () => {
  beforeEach(() => {
    resetOut();
  });

  it("initFlow with appId", () => {
    MaestroTranslators.initFlow({ appId: "testAppId" });
    expect(getOutput()).toMatchSnapshot();
  });

  it("launchApp with appId", () => {
    MaestroTranslators.launchApp({ appId: "testAppId" });
    expect(getOutput()).toMatchSnapshot();
  });

  it("clearState with appId", () => {
    MaestroTranslators.clearState({ appId: "testAppId" });
    expect(getOutput()).toMatchSnapshot();
  });

  it("clearKeychain", () => {
    MaestroTranslators.clearKeychain();
    expect(getOutput()).toMatchSnapshot();
  });
});
