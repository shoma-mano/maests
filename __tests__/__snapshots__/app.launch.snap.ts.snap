// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MaestroTranslators Application Launch Commands Snapshots > clearKeychain 1`] = `
"- clearKeychain
"
`;

exports[`MaestroTranslators Application Launch Commands Snapshots > clearState with appId 1`] = `
"- clearState: testAppId
"
`;

exports[`MaestroTranslators Application Launch Commands Snapshots > initFlow with appId 1`] = `
"appId: testAppId
---
"
`;

exports[`MaestroTranslators Application Launch Commands Snapshots > launchApp with appId 1`] = `
"- launchApp:
    appId: "testAppId"
"
`;
